### ISSUE TITLE
cmd/go: support corpus minimization

### ISSUE THREAD

**Issue by katiehockman** [golang/go#49290]
We should support corpus minimization, which will mean the ability to remove items from the on-disk corpus which either 1) don't have the types that are supported by the fuzz test (e.g. they are left over from a previous version of the test which took different params), or 2) don't expand any new coverage that isn't already provided by other entries in the corpus.

(1) should be pretty straightforward. We can just unmarshal the contents of the file, and see if it matches. If it doesn't, delete it.

(2) will be a bit more involved, but not necessarily that complicated. We should take a look at how libFuzzer implements this. One potential solution would be to maintain a coverage map, and run each corpus item against the fuzz test in turn, updating the map is it runs. If any of them don't expand coverage, delete it. A potential pitfall of this: if we have 20 corpus entries that each expand 1 line, and 1 corpus entry that covers all 20 of those lines at once, which do we choose?

At least (2) will be needed for OSS-Fuzz integration.

**Comment by bcmills** [comment]
A thought regarding (2): if we have an effective minimization algorithm, is it actually necessary to prune the corpus on disk? It probably is worthwhile to prune the inputs stored in the Go build cache, but arguably the former-crashers in the `testdata/fuzz` directory should be retained as regression tests.

Perhaps we could instead start each fuzzing run by spending some fraction of the time budget re-minimizing the corpus. That would also avoid losing coverage if (for example) the code is modified in a way that significantly changes coverage before refactoring back to something closer to a previous approach (for which the pruned-out inputs might actually become relevant again).

**Comment by katiehockman** [comment]
> It probably is worthwhile to prune the inputs stored in the Go build cache, but arguably the former-crashers in the testdata/fuzz directory should be retained as regression tests.

Yes agreed. When I talk about "corpus minimization" here, I'm only referring to the corpus in the build cache. We shouldn't touch the seed corpus.

> Perhaps we could instead start each fuzzing run by spending some fraction of the time budget re-minimizing the corpus.

We could do this, but I would argue it should be opt in. e.g. `go test -fuzz=Fuzz -fuzzminimizecorpus`
Otherwise, I'm imagining a scenario where someone has a fuzz target that accepts `[]byte`. In playing around with their test, they change the target to accept `[]byte, int`. If we prune the corpus by default, then when they run `go test -fuzz Fuzz`, all of the cache corpus that only accepts `[]byte` is going to be deleted before the user realizes what happened.

**Comment by katiehockman** [comment]
@dgryski pointed me to https://arxiv.org/pdf/1905.13055.pdf. Commenting here so we can find it later.

**Comment by ianlancetaylor** [comment]
Moving to Backlog.

**Comment by morehouse** [comment]
Bump.

Because corpus minimization/merging features are missing, it is difficult for major Go projects to maintain public corpora.  PRs adding new inputs to the corpus can't be easily evaluated to determine which seeds actually increase coverage, and we essentially have to accept all new inputs as a package based on coverprofiles before/after the PR.  Over time, this can lead to major bloat of the corpora.

Those serious about fuzzing Go end up resorting to libFuzzer build mode and doing [hacky things](https://github.com/google/oss-fuzz/blob/master/infra/base-images/base-runner/gocoverage/convertcorpus/main.go#L38) to measure code coverage.

It would be so much nicer if we could just use the native fuzzing tools.

**Comment by gijswijs** [comment]
Let me bump this as well, to see if this can get some traction.

**Comment by morehouse** [comment]
I ended up writing a hacky [script](https://github.com/lightninglabs/lnd-fuzz/blob/1b37a8bcfb2f3c02d436fec7a097e731fe2c80bc/corpus_merge.sh#L53) to do corpus minimization, since this issue hasn't gotten any traction.

The script sets `GODEBUG="fuzzdebug=1"` and then reads debug output during initial corpus processing to extract the number of coverage bits.  The script is quite slow and inefficient (`O(n^2)`) since coverage bits must be extracted out-of-process, and so every time a new input is selected to be included in the minimal corpus, the full updated corpus must be processed again.

A native minimization tool for Go would be more reliable long term and much faster (`O(n)`), since coverage bits could be measured in-process and accumulated in a bitmap to determine which inputs to merge.  See libFuzzer's [merge logic](https://github.com/llvm/llvm-project/blob/9e33cb22f991fb25d606d89c0e5a13a3ebed52fe/compiler-rt/lib/fuzzer/FuzzerMerge.cpp#L139-L194) for inspiration.

**Comment by aaron42net** [comment]
I played around in this space and ended up with [gofuzzmin](https://github.com/die-net/gofuzzmin), a tool which removes cache seeds that don't expand coverage.

It builds the test binary with the fuzzer's own edge-counter instrumentation (-gcflags="-d=libfuzzer"), sorts cache seeds shortest-first, then runs a single warmup phase with GODEBUG="fuzzdebug=1" to parse the new bits each seed contributes. Seeds that add zero new coverage are redundant and can be deleted with -prune. This uses the same 8-bit edge counters the fuzzer uses internally, so the minimized corpus preserves exactly the coverage the fuzzer cares about.

This prefers keeping whichever seeds it processes first that increase coverage and currently only sorts the seeds based on length.  afl-cmin seems to prefer faster-to-run seeds, with a sub-sort on length.  So there's maybe room for improvement that might just involve adding more information to the sort.

Building a test binary, running a single warmup phase, and parsing the output turns out to be pretty fast.  With a warmed-up compiler for the test binary, the whole thing is done in a few seconds with a cache containing a few thousand seeds.

This obviously reaches into the guts of the fuzzer implementation and may break with any significant refactoring there.
